### ISSUE TITLE
Updating a running sequence in the timeline

### ISSUE THREAD

**Issue by woudsma** [ideoforms/isobar#47]
First of all, thanks for this project! I'm working on a Raspberry Pi based MIDI sequencer and I'm using `isobar` to sync the timeline to an external MIDI clock, and to send MIDI events to external gear (I'm testing with a MFB-503 drumcomputer).

When I press a button on the hardware sequencer, a Node.js script sends the updated sequence data into the Python process where `isobar` is running.

I just can't seem to figure out how to update a running sequence without it falling silent until it repeats.
I want to update the notes in place, basically swapping notes/CC in the existing sequence while it keeps playing/repeating.

In the picture I've got 3 tracks (16, 14, 10 steps) with some trigs as an example, each track is scheduled separately in `isobar`.
                                                                                                                                                                                                                                                                                                                           
![pi-sequencer](https://github.com/ideoforms/isobar/assets/6162978/6f090d95-46d4-43c7-ae75-040324810d92)

It kind of works when I set the `quantize=4` parameter. The tracks will sync up correctly when the pattern loops after 4 beats (16 steps with a duration of 0.25). But if the sequence is running and I update it when e.g. 10 steps are remaining, the final 10 steps are silent until the pattern loops again.

This is my (simplified) current approach:
```py
_, bpm, midi_clock_in_name, midi_mfb503_out_name = sys.argv

midi_in = iso.MidiInputDevice(device_name=midi_clock_in_name)
midi_out = iso.MidiOutputDevice(device_name=midi_mfb503_out_name, send_clock=True)
timeline = iso.Timeline(bpm, clock_source=midi_in, output_device=midi_out)

events = {}
tracks = {}

while True:
  try:
    line = input().strip() # receive data from parent process on sequence update
    id, trigs, velocities, notes = line.split()

    track_name = f"track-{id}"
    note_seq, velocity_seq = get_seq(trigs, velocities, notes)

    # updated sequence:
    # note_seq:     [48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48, 48]
    # velocity_seq: [127, 0, 0, 0, 127, 0, 0, 0, 127, 0, 127, 0, 127, 0, 0, 0]

    events[track_name] = {
      iso.EVENT_NOTE: iso.PSequence(note_seq),
      iso.EVENT_AMPLITUDE: iso.PSequence(velocity_seq),
      iso.EVENT_DURATION: 0.25,
    }

    if track_name in tracks:
      tracks[track_name].update(events[track_name], quantize=4)
    else:
      tracks[track_name] = timeline.schedule(params=events[track_name], name=track_name, quantize=4)

  except EOFError:
    break

timeline.run()
```
How can I update the sequence without it stopping for a few beats? I want the current pattern to keep playing, while updating one or more note/CC values in the pattern.

Any thoughts/tips are much appreciated!

**Comment by ideoforms** [comment]
Thanks for the report. Love the look of the sequencer, and great choice of drum machines - I have a couple of MFB drum synths (301 and Tanzbar), can't get enough of their characterful sound!

I've just looked into this and it looks like a bug; the existing pattern should definitely continue playing until the scheduled time of the `update`. Will see about implementing a fix now.

**Comment by ideoforms** [comment]
@woudsma I've just tested and pushed an update that should fix the problem. Please clone and install the latest commit, and let me know if that resolves the issue for you. Thanks!

**Comment by woudsma** [comment]
@ideoforms Thanks for your reply! Yes the MFB machines sound amazing indeed, digital drumcomputers such as Digitakt or TR-8(S) don't come close IMHO. I do find the Electron-style sequencing much better than most analog drumcomputer built-in sequencers though.

I'm glad you like the look of the sequencer! Everything is still far from perfect but the current UI is built with Electron + React. It's not as performant as `node-raylib` for example, but developing/styling in a browser with Chrome DevTools is much faster. Fortunately the latency is quite minimal since `isobar` runs in it's own process. The app is running on a RPi 4B with a cheap 4.3" DSI Touchscreen and custom PCB. The LED/switches/encoders GPIO events are handled by a Python script and another Python script handles the MIDI IO using `isobar`. It is the only Python library that I could find that could properly sync to an external MIDI clock, amazing work 👍 

I've just installed your fix and it seems to be working! The sequence keeps playing and it updates after the pattern repeats.

I did notice that some sequence updates won't come through. It keeps playing a note when that note has been disabled for example. Maybe this has to do something with updating my sequence multiple times before it repeats?
```py
# currently playing (1=on, 0=off)
1000100010001000
# updated sequence
1110100010001000

# plays
1100100010001000
```
It seems like it sometimes doesn't update the sequence. If I only do a single sequence update per 4 beats (so if I wait until the pattern repeats before doing another update) everything seems to work as intended. I get the same result with and without the `replace=True` parameter.


https://github.com/ideoforms/isobar/assets/6162978/ec39810a-d65a-42b4-ab65-a331e3a4c75f

**Comment by woudsma** [comment]
I could probably get away with queuing the sequence updates so that only one update is performed right before the pattern repeats. Maybe a bit hacky but it might be worth a try.

**Comment by ideoforms** [comment]
Ah, oops, I didn’t handle the case of multiple updates in the same interval. Will address - longer reply coming soon!

**Comment by ideoforms** [comment]
OK, just pushed a fix and test for this - can you pull and test again?

@woudsma In reply to your previous comments: completely agree about the Elektron sequencing style, it's so much more immediate and intuitive. I love the Tanzbär's sound, but the sequencer is extremely idiosyncratic to say the least :) And on the more UI-limited MFB devices without a proper step sequencer, I don't bother with it at all. I like using the Digitakt as a sequencer for the 301 (although I wish it were possible to store groups of MIDI CC presets, as I still need to reach for the manual to remember which CC# controls which drum parameter....)

The video you shared is just awesome. Would really like to play with this! What is the keyboard-style controller that you're using? Is this something custom built?

**Comment by woudsma** [comment]
@ideoforms You're a hero! The fix works like a charm.

Yeah the Digitakt sequencer is truly amazing, I've used up all my Digitakt MIDI tracks for the Alpha Base (which also has a horrible UI / sequencer but sounds great). I wish Electron would release a standalone sequencer that isn't limited by 8 (or 12 on the Syntakt) MIDI tracks. More LFO's per MIDI track would be nice as well. I was/am looking for a Cirklon Sequentix but they are hard to find for a reasonable price, also their waiting list for a new one is ~4 years.. I thought I could make something less expensive in less time than that 😅 

The controller is a custom PCB made with KiCad and I've found the components in various webshops. It's wired up to the RPi GPIO pins. This is my first hardware project working with components/soldering and ChatGPT has been helpful with figuring out some things. I can't believe it actually works but it does! There are still some bugs and reading the correct hardware values is challenging. Sometimes it doesn't register a switch press or encoder rotation for example. Luckily performance is not that much of an issue because I don't have to process audio. I've got a few [pictures of the project here](https://www.tumblr.com/wousma/721573983597756417/heron-hardware-midi-sequencer-work-in-progress?source=share). Maybe I'll try to build a 16-step version in the future if I can iron out some of the bugs, haha. Some more buttons would be nice.

First I'm trying to finish trig conditions, next maybe some sort of LFO for CC values. I'll have to think about that though. Have you ever tried to implement something like an LFO with `isobar`? I'd be curious if you have, but if not I'll play around with the extensive sequencing operators anyway! Thanks again for your quick responses and fixes 👌
